File: packages/provide-uterm-go/gateway/iac.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

// Package gateway ports the reverse-direction gateway classes of provide-uterm
// (provide.uterm.gateway) to Go: a raw-TCP (telnet) listener and an SSH server
// that each accept inbound clients and proxy every byte to an upstream
// WebSocket terminal server through the inline DLE/STX control channel.
//
// It is the mirror image of the browser-facing proxy: instead of a browser
// WebSocket dialing out to telnet/SSH, traditional telnet/SSH clients dial in
// and are bridged out to a WebSocket. See TelnetWsGateway and SshWsGateway.
package gateway

import "strings"

// Telnet byte vocabulary (RFC 854 + option codes). Mirrors _iac_negotiate.py.
const (
	iacIAC  = 255
	iacDONT = 254
	iacDO   = 253
	iacWONT = 252
	iacWILL = 251
	iacSB   = 250
	iacSE   = 240

	iacBREAK = 243
	iacIP    = 244
	iacEOF   = 236

	optTTYPE      = 24 // RFC 1091
	optNewEnviron = 39 // RFC 1572

	subIS      = 0
	subSEND    = 1
	envVar     = 0
	envValue   = 1
	envEsc     = 2
	envUserVar = 3

	// maxSBBytes bounds a single IAC subnegotiation so a hostile client that
	// opens IAC SB and never sends IAC SE cannot grow the buffer without limit.
	maxSBBytes = 4096
)

// DeriveColormode picks the best ?colormode= value from a TERM + env map.
//
// Precedence (first match wins; "" when nothing applies):
//  1. COLORTERM == truecolor / 24bit → passthrough
//  2. TERM ending in -direct / -truecolor → passthrough
//  3. TERM ending in -256color → 256
//  4. Legacy TERM (xterm / vt100 / …) → 16
//
// Case-insensitive; a missing term falls back to env["TERM"].
func DeriveColormode(term string, env map[string]string) string {
	colorterm := strings.ToLower(strings.TrimSpace(env["COLORTERM"]))
	if colorterm == "truecolor" || colorterm == "24bit" {
		return "passthrough"
	}
	t := term
	if t == "" {
		t = env["TERM"]
	}
	t = strings.ToLower(strings.TrimSpace(t))
	if strings.HasSuffix(t, "-direct") || strings.HasSuffix(t, "-truecolor") || t == "xterm-direct" {
		return "passthrough"
	}
	if strings.HasSuffix(t, "-256color") || t == "xterm-256color" {
		return "256"
	}
	switch t {
	case "xterm", "vt100", "vt102", "vt220", "ansi", "linux", "dumb":
		return "16"
	}
	return ""
}

// stripIAC removes IAC telnet negotiation sequences from inbound client data,
// translating IP/BREAK to Ctrl-C and EOF to Ctrl-D. Stateless fallback used
// when IAC negotiation is disabled (mirrors _strip_iac).
func stripIAC(data []byte) []byte {
	out := make([]byte, 0, len(data))
	i, n := 0, len(data)
	for i < n {
		b := data[i]
		if b != iacIAC {
			out = append(out, b)
			i++
			continue
		}
		if i+1 >= n {
			break
		}
		cmd := data[i+1]
		switch cmd {
		case iacIAC:
			out = append(out, iacIAC)
			i += 2
		case iacSB:
			i = skipSubneg(data, i+2, n)
		case iacIP, iacBREAK:
			out = append(out, 0x03)
			i += 2
		case iacEOF:
			out = append(out, 0x04)
			i += 2
		case iacWILL, iacWONT, iacDO, iacDONT:
			if i+2 >= n {
				return out
			}
			i += 3
		default:
			i += 2
		}
	}
	return out
}

// skipSubneg scans forward from i (first byte after IAC SB) to just past the
// closing IAC SE pair, or n if the sequence is truncated.
func skipSubneg(data []byte, i, n int) int {
	for i < n {
		if data[i] == iacIAC && i+1 < n && data[i+1] == iacSE {
			return i + 2
		}
		i++
	}
	return n
}

// IacNegotiator is a stateful IAC negotiator: it reads client bytes, emits
// replies, and collects TERM/env hints. Single-use per TCP connection. Direct
// port of the Python IacNegotiator.
type IacNegotiator struct {
	Term string
	Env  map[string]string

	sbOption   int // -1 when not inside a subnegotiation
	sbBuf      []byte
	sbOverflow bool
	pending    []byte

	ttypeRequested      bool
	newEnvironRequested bool
	ttypeReceived       bool
	newEnvironReceived  bool
}

// NewIacNegotiator returns a fresh negotiator.
func NewIacNegotiator() *IacNegotiator {
	return &IacNegotiator{Env: map[string]string{}, sbOption: -1}
}

// StartBytes returns the initial IAC bytes to send on session start
// (IAC DO TTYPE, IAC DO NEW-ENVIRON).
func (g *IacNegotiator) StartBytes() []byte {
	g.ttypeRequested = true
	g.newEnvironRequested = true
	return []byte{iacIAC, iacDO, optTTYPE, iacIAC, iacDO, optNewEnviron}
}

// Feed consumes data from the client and returns (reply, cleaned): reply is
// bytes to echo back to the client, cleaned is the application data with IAC
// noise removed.
func (g *IacNegotiator) Feed(data []byte) (reply, cleaned []byte) {
	if len(g.pending) > 0 {
		data = append(append([]byte(nil), g.pending...), data...)
		g.pending = nil
	}
	var rep, clean []byte
	i, n := 0, len(data)
	for i < n {
		if g.sbOption >= 0 {
			if data[i] == iacIAC && i+1 < n && data[i+1] == iacSE {
				g.finishSB()
				i += 2
				continue
			}
			if data[i] == iacIAC && i+1 < n && data[i+1] == iacIAC {
				g.appendSB(iacIAC)
				i += 2
				continue
			}
			g.appendSB(data[i])
			i++
			continue
		}
		b := data[i]
		if b != iacIAC {
			clean = append(clean, b)
			i++
			continue
		}
		if i+1 >= n {
			g.pending = append([]byte(nil), data[i:]...)
			break
		}
		cmd := data[i+1]
		switch cmd {
		case iacIAC:
			clean = append(clean, iacIAC)
			i += 2
		case iacSB:
			if i+2 >= n {
				g.pending = append([]byte(nil), data[i:]...)
				return rep, clean
			}
			g.sbOption = int(data[i+2])
			g.sbBuf = nil
			g.sbOverflow = false
			i += 3
		case iacWILL, iacWONT, iacDO, iacDONT:
			if i+2 >= n {
				g.pending = append([]byte(nil), data[i:]...)
				return rep, clean
			}
			rep = append(rep, g.handleOption(cmd, data[i+2])...)
			i += 3
		default:
			i += 2
		}
	}
	return rep, clean
}

// Done reports whether every requested option has been answered.
func (g *IacNegotiator) Done() bool {
	ttypeOK := !g.ttypeRequested || g.ttypeReceived
	envOK := !g.newEnvironRequested || g.newEnvironReceived
	return ttypeOK && envOK
}

// DerivedColormode maps the captured hints to a ?colormode= value.
func (g *IacNegotiator) DerivedColormode() string { return DeriveColormode(g.Term, g.Env) }

func (g *IacNegotiator) handleOption(verb, option byte) []byte {
	if verb == iacWILL && option == optTTYPE {
		return []byte{iacIAC, iacSB, optTTYPE, subSEND, iacIAC, iacSE}
	}
	if verb == iacWILL && option == optNewEnviron {
		return []byte{iacIAC, iacSB, optNewEnviron, subSEND, iacIAC, iacSE}
	}
	return nil
}

func (g *IacNegotiator) appendSB(b byte) {
	if len(g.sbBuf) >= maxSBBytes {
		g.sbOverflow = true
		return
	}
	g.sbBuf = append(g.sbBuf, b)
}

func (g *IacNegotiator) finishSB() {
	option := g.sbOption
	payload := g.sbBuf
	overflowed := g.sbOverflow
	g.sbOption = -1
	g.sbBuf = nil
	g.sbOverflow = false
	if overflowed {
		return
	}
	switch option {
	case optTTYPE:
		g.Term = parseTTypeIS(payload)
		g.ttypeReceived = true
	case optNewEnviron:
		g.Env = parseNewEnvironIS(payload)
		g.newEnvironReceived = true
	}
}

// parseTTypeIS extracts the lowercased terminal name from a TTYPE IS payload.
func parseTTypeIS(payload []byte) string {
	if len(payload) == 0 || payload[0] != subIS {
		return ""
	}
	return strings.ToLower(strings.TrimSpace(string(payload[1:])))
}

// parseNewEnvironIS parses an RFC 1572 NEW-ENVIRON IS payload into a map.
func parseNewEnvironIS(payload []byte) map[string]string {
	out := map[string]string{}
	if len(payload) == 0 || payload[0] != subIS {
		return out
	}
	i, n := 1, len(payload)
	for i < n {
		marker := payload[i]
		if marker != envVar && marker != envUserVar {
			return out
		}
		i++
		var name []byte
		for i < n && payload[i] != envValue && payload[i] != envVar && payload[i] != envUserVar {
			if payload[i] == envEsc && i+1 < n {
				name = append(name, payload[i+1])
				i += 2
				continue
			}
			name = append(name, payload[i])
			i++
		}
		var value []byte
		if i < n && payload[i] == envValue {
			i++
			for i < n && payload[i] != envVar && payload[i] != envUserVar {
				if payload[i] == envEsc && i+1 < n {
					value = append(value, payload[i+1])
					i += 2
					continue
				}
				value = append(value, payload[i])
				i++
			}
		}
		nm := strings.TrimSpace(string(name))
		if nm != "" {
			out[nm] = string(value)
		}
	}
	return out
}

File: packages/provide-uterm-ts/src/gateway/iac.ts
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

/**
 * Working out what kind of terminal is at the far end of a telnet connection.
 *
 * Port of `provide.uterm.gateway._iac_negotiate`. Before anything useful
 * crosses, the gateway asks the client two questions — what terminal it is,
 * and what its environment says — and the answers decide the colour mode the
 * upstream session is opened with.
 *
 * * **Every IAC byte comes out of the stream.** What the client typed is what
 *   goes upstream; a stray `IAC` reaching a shell is a byte nobody typed.
 * * **A subnegotiation can arrive in pieces**, since this is a byte stream, so
 *   what is incomplete is held and carried on with.
 * * **The buffer is capped**, so a client that opens a subnegotiation and
 *   never closes it costs a fixed amount of memory rather than all of it.
 */

/** The telnet vocabulary this needs (RFC 854 and the option codes). */
const IAC = 255;
const DONT = 254;
const DO = 253;
const WONT = 252;
const WILL = 251;
const SB = 250;
const SE = 240;

/** Terminal type (RFC 1091). */
const OPT_TTYPE = 24;
/** Environment (RFC 1572). */
const OPT_NEW_ENVIRON = 39;

const SUB_IS = 0;
const SUB_SEND = 1;
const ENV_VAR = 0;
const ENV_VALUE = 1;
const ENV_ESC = 2;
const ENV_USERVAR = 3;

/** Terminals whose colour support is known without asking. */
const LEGACY_TERMINALS = new Set(["xterm", "vt100", "vt102", "vt220", "ansi", "linux", "dumb"]);

/** How much of an unfinished subnegotiation is kept. */
export const MAX_SUBNEGOTIATION_BYTES = 512;

/** What a client says back, and what goes on upstream. */
export interface IacFeedResult {
  /** Bytes to write back to the client. */
  reply: Uint8Array;
  /** The client's own data, with every IAC sequence taken out. */
  cleaned: Uint8Array;
}

/** `IAC SB <option> SEND IAC SE` — asking a client to send something. */
function askFor(option: number): Uint8Array {
  return Uint8Array.from([IAC, SB, option, SUB_SEND, IAC, SE]);
}

/**
 * The colour mode a terminal and environment imply.
 *
 * A true-colour hint wins outright; otherwise a terminal naming 256 colours
 * gets 256, and one this recognises at all gets 16. A terminal nobody
 * recognises gets nothing, so the upstream is left to decide rather than
 * being told something wrong.
 */
export function deriveColormode(
  term: string | undefined,
  env: Readonly<Record<string, string>> = {},
): string | undefined {
  const colorterm = (env.COLORTERM ?? "").trim().toLowerCase();
  if (colorterm === "truecolor" || colorterm === "24bit") {
    return "passthrough";
  }
  // A terminal named by the client wins; failing that, one it put in its
  // environment.
  const name = (term === undefined || term === "" ? (env.TERM ?? "") : term).trim().toLowerCase();
  if (name.endsWith("-direct") || name.endsWith("-truecolor")) {
    return "passthrough";
  }
  if (name.endsWith("-256color")) {
    return "256";
  }
  // Named exactly, not by prefix: `xterm-color` is not one of these, and
  // guessing sixteen colours for a terminal nobody listed would be telling
  // the upstream something that might be wrong.
  if (LEGACY_TERMINALS.has(name)) {
    return "16";
  }
  return undefined;
}

/** Follows the handshake far enough to learn what the client is. */
export class IacNegotiator {
  /** What the client said it is. */
  term = "";
  /** What the client said about its environment. */
  readonly env: Record<string, string> = {};

  #pending: number[] = [];
  #inSubnegotiation = false;
  #subnegotiation: number[] = [];
  #ttypeRequested = false;
  #newEnvironRequested = false;
  #ttypeReceived = false;
  #newEnvironReceived = false;

  /** What to send the moment a client connects. */
  startBytes(): Uint8Array {
    this.#ttypeRequested = true;
    this.#newEnvironRequested = true;
    return Uint8Array.from([IAC, DO, OPT_TTYPE, IAC, DO, OPT_NEW_ENVIRON]);
  }

  /** Whether every question asked has been answered. */
  done(): boolean {
    const ttypeOk = !this.#ttypeRequested || this.#ttypeReceived;
    const environOk = !this.#newEnvironRequested || this.#newEnvironReceived;
    return ttypeOk && environOk;
  }

  /** The colour mode implied by what the client has said so far. */
  derivedColormode(): string | undefined {
    return deriveColormode(this.term, this.env);
  }

  /**
   * Consume bytes from the client.
   *
   * Anything left half-finished is held for the next call, so a sequence split
   * across two reads is completed rather than lost.
   */
  feed(data: Uint8Array): IacFeedResult {
    const reply: number[] = [];
    const cleaned: number[] = [];
    // Whatever was held back last time goes in front of what has just come.
    const bytes = this.#pending.length === 0 ? [...data] : [...this.#pending, ...data];
    this.#pending = [];

    let index = 0;
    while (index < bytes.length) {
      const byte = bytes[index] as number;

      if (this.#inSubnegotiation) {
        index = this.#consumeSubnegotiation(bytes, index);
        continue;
      }

      if (byte !== IAC) {
        cleaned.push(byte);
        index += 1;
        continue;
      }

      if (index + 1 >= bytes.length) {
        // A lone IAC at the end: the rest is still on its way.
        this.#pending = bytes.slice(index);
        break;
      }

      const command = bytes[index + 1] as number;
      if (command === IAC) {
        // An escaped IAC is one literal byte of the client's own data.
        cleaned.push(IAC);
        index += 2;
        continue;
      }
      if (command === SB) {
        this.#inSubnegotiation = true;
        // Cleared here as well as on completion: finishing one empties it, so
        // this only matters if that ever stops being true.
        this.#subnegotiation = [];
        index += 2;
        continue;
      }
      if (command === WILL || command === WONT || command === DO || command === DONT) {
        if (index + 2 >= bytes.length) {
          // The option byte has not arrived yet.
          this.#pending = bytes.slice(index);
          break;
        }
        reply.push(...this.#handleOption(command, bytes[index + 2] as number));
        index += 3;
        continue;
      }
      // A control the outer gateway already handles. Not application data.
      index += 2;
    }

    return { reply: Uint8Array.from(reply), cleaned: Uint8Array.from(cleaned) };
  }

  /** Read a subnegotiation up to its terminator, if it is all here. */
  #consumeSubnegotiation(bytes: readonly number[], start: number): number {
    let index = start;
    while (index < bytes.length) {
      const byte = bytes[index] as number;
      if (byte !== IAC) {
        this.#append(byte);
        index += 1;
        continue;
      }
      if (index + 1 >= bytes.length) {
        // The byte after IAC decides whether this ends; wait for it.
        this.#pending = bytes.slice(index);
        return bytes.length;
      }
      const next = bytes[index + 1] as number;
      if (next === IAC) {
        this.#append(IAC);
        index += 2;
        continue;
      }
      if (next === SE) {
        this.#inSubnegotiation = false;
        this.#finish();
        return index + 2;
      }
      // An IAC that is neither escaped nor a terminator: not part of the
      // payload, and not something to act on inside one.
      index += 2;
    }
    return index;
  }

  /** Keep a subnegotiation byte, up to the cap. */
  #append(byte: number): void {
    if (this.#subnegotiation.length < MAX_SUBNEGOTIATION_BYTES) {
      // Past the cap the bytes are dropped rather than the connection: a
      // client that never closes one costs a fixed amount either way.
      this.#subnegotiation.push(byte);
    }
  }

  /** Read what a completed subnegotiation said. */
  #finish(): void {
    const payload = this.#subnegotiation;
    this.#subnegotiation = [];
    if (payload.length < 2) {
      // Stated rather than left to the `kind` check below, which would also
      // refuse it — reading past the end happens to give `undefined` here,
      // and that is not a thing to rely on.
      return;
    }
    const option = payload[0] as number;
    const kind = payload[1] as number;
    if (kind !== SUB_IS) {
      return;
    }
    const body = payload.slice(2);
    if (option === OPT_TTYPE) {
      this.#ttypeReceived = true;
      // Lowercased and trimmed on the way in, as the reference stores it, so
      // a client shouting its terminal type is the same client.
      this.term = String.fromCharCode(...body)
        .trim()
        .toLowerCase();
      return;
    }
    if (option === OPT_NEW_ENVIRON) {
      this.#newEnvironReceived = true;
      this.#readEnvironment(body);
    }
  }

  /** Read the variable/value pairs out of an environment subnegotiation. */
  #readEnvironment(body: readonly number[]): void {
    let key: number[] = [];
    let value: number[] = [];
    // Nothing is a variable until a marker says one has started, so bytes
    // before the first marker belong to nobody.
    let started = false;
    let reading: "key" | "value" = "key";

    const store = (): void => {
      if (started && key.length > 0) {
        this.env[String.fromCharCode(...key)] = String.fromCharCode(...value);
      }
    };

    for (let index = 0; index < body.length; index += 1) {
      const byte = body[index] as number;
      if (byte === ENV_VAR || byte === ENV_USERVAR) {
        store();
        key = [];
        value = [];
        started = true;
        reading = "key";
        continue;
      }
      if (byte === ENV_VALUE) {
        reading = "value";
        continue;
      }
      if (byte === ENV_ESC) {
        // The next byte is literal, whatever it would otherwise mean.
        index += 1;
        if (index >= body.length) {
          break;
        }
      }
      const literal = body[index] as number;
      if (reading === "key") {
        key.push(literal);
      } else {
        value.push(literal);
      }
    }
    store();
  }

  /** Answer a verb about an option. */
  #handleOption(verb: number, option: number): number[] {
    // A client accepting one of the two questions is asked to answer it.
    if (verb === WILL && option === OPT_TTYPE) {
      return [...askFor(OPT_TTYPE)];
    }
    if (verb === WILL && option === OPT_NEW_ENVIRON) {
      return [...askFor(OPT_NEW_ENVIRON)];
    }
    // Everything else gets silence — including a client offering something
    // nobody asked for, which RFC 854 says should be refused. Carried over
    // from the reference; recorded in the roadmap rather than corrected,
    // since a port that answered differently would negotiate differently.
    return [];
  }
}
